src/parser.cpp

```diff
@@ -1503,10 +1503,13 @@ on_headers(
                 return;
             }
 
+            BOOST_ASSERT(fb_.max_size() >= h_.size);
+
             auto n0 =
-                fb_.capacity() - h_.size +
+                fb_.max_size() - h_.size +
```

Comment:
`fb_.capacity()` is used in a few other places too.
Should we assume we will always call `shrink_to_fit` before calling `prepare` on `fb_`?"

Reply:
No do not assume that

Reply:
I'm not convinced that `flat_buffer` even has a design flaw, being honest.

We should save any calls of `shrink_to_fit()` once we have a concrete problem and therefore a use-case.

Reply:
Without knowing where we are going to call `shrink_to_fit`, it is hard to judge if using `max_size` instead of `capacity` is appropriate, because it only works if we guarantee:

- never reach the `max_size`
- completely consume the buffer before the next call to `prepare`

> We should save any calls of shrink_to_fit() once we have a concrete problem and therefore a use-case.

An automatic `memmove` in `flat_buffer::prepare` would only happen if the requested size requires it (otherwise, `prepare` would throw an exception).

Reply:
No one is contemplating calling `shrink_to_fit` from within `flat_buffer`. The function `shrink_to_fit` as of now is ONLY designed as an available API for users to invoke. Automating shrink to fit is a new feature which we 1. do not have a use-case for, and 2. are not at the point in our development cycle where we can answer the necessary design questions which arise from it.